packages/@lwc/node-reactions/src/traverse.ts

```diff
@@ -31,12 +35,15 @@ export default function queueReactionsForTree(
         });
     }
 
-    if (root instanceof Element && !!root.shadowRoot) {
+    const nodeType = nodeTypeGetter.call(root);
+    // Perf optimization: Only Element type can have shadowRoot
+    if (nodeType === 1 && (root as Element).shadowRoot) {
         // intentionally co-oercing to a truthy value because shadowRoot property can be undefined or null(HTMLUnknownElement)
-        queueReactionsForTree(root.shadowRoot, reactionTypes, reactionQueue);
+        queueReactionsForTree((root as Element).shadowRoot!, reactionTypes, reactionQueue);
     }
 
-    if (isTrue(hasChildNodes.call(root))) {
+    // Perf optimization: Element and Document Fragment are the only types that need to be processed
+    if ((nodeType === 1 || nodeType === 11) && isTrue(hasChildNodes.call(root))) {
```

Comment:
probably the call to hasChildNodes then childNodesGetter is slower than just childNodesGetter because you cross the bridge only once.